bin/light-base/src/sync_service/parachain.rs

```diff
@@ -45,752 +45,1032 @@ pub(super) async fn start_parachain<TPlat: Platform>(
-                        }
-                        ToBackground::PeersAssumedKnowBlock { send_back, block_number, block_hash } => {
-                            // If `block_number` is over the finalized block, then which source
-                            // knows which block is precisely tracked. Otherwise, it is assumed
-                            // that all sources are on the finalized chain and thus that all
-                            // sources whose best block is superior to `block_number` have it.
-                            let list = if block_number > sync_sources.finalized_block_height() {
-                                sync_sources.knows_non_finalized_block(block_number, &block_hash)
-                                    .map(|local_id| sync_sources[local_id].0.clone())
-                                    .collect()
-                            } else {
-                                sync_sources
-                                    .keys()
-                                    .filter(|local_id| {
-                                        sync_sources.best_block(*local_id).0 >= block_number
-                                    })
-                                    .map(|local_id| sync_sources[local_id].0.clone())
-                                    .collect()
+
+                        foreground_message = self.from_foreground.next().fuse() => {
+                            // Message from the public API of the syncing service.
+
+                            // Terminating the parachain sync task if the foreground has closed.
+                            let foreground_message = match foreground_message {
+                                Some(m) => m,
+                                None => return,
                             };
 
-                            let _ = send_back.send(list);
-                        }
-                        ToBackground::SyncingPeers { send_back } => {
-                            let _ = send_back.send(sync_sources.keys().map(|local_id| {
-                                let (height, hash) = sync_sources.best_block(local_id);
-                                let (peer_id, role) = sync_sources[local_id].clone();
-                                (peer_id, role, height, *hash)
-                            }).collect());
+                            self.process_foreground_message(foreground_message).await;
+                        },
+                    }
+                }
+
+                ParachainBackgroundState::Subscribed(ref mut runtime_subscription) => {
+                    futures::select! {
+                        () = &mut runtime_subscription.next_start_parahead_fetch => {
+                            // Do nothing. This is simply to wake up and loop again.
+                        },
+
+                        relay_chain_notif = runtime_subscription.relay_chain_subscribe_all.next().fuse() => { // TODO: remove fuse()?
```

Comment:
extract into `process_relay_chain_notification`?

Reply:
I don't think it's a good idea.
The block that I've left it handles situations where the channel of relay chain notifications has been closed. In other words, it's more "meta" than processing the notification itself.